core/txpool/blobpool/blobpool.go

```diff
@@ -0,0 +1,1461 @@
+		return
+	}
+	block, ok := inclusions[item.Tx.Hash()]
+	if !ok {
+		log.Warn("Blob transaction swapped out by signer", "from", addr, "nonce", nonce, "id", id)
+		return
+	}
+	if err := p.limbo.push(item.Tx.Hash(), block, item.Blobs, item.Commits, item.Proofs); err != nil {
+		log.Warn("Failed to offload blob tx into limbo", "err", err)
+		return
+	}
+}
+
+// Reset implements txpool.SubPool, allowing the blob pool's internal state to be
+// kept in sync with the main transacion pool's internal state.
+func (p *BlobPool) Reset(oldHead, newHead *types.Header) {
+	waitStart := time.Now()
+	p.lock.Lock()
+	resetwaitHist.Update(time.Since(waitStart).Nanoseconds())
+	defer p.lock.Unlock()
+
+	defer func(start time.Time) {
+		resettimeHist.Update(time.Since(start).Nanoseconds())
+	}(time.Now())
+
+	statedb, err := p.chain.StateAt(newHead.Root)
+	if err != nil {
+		log.Error("Failed to reset blobpool state", "err", err)
+		return
+	}
+	p.head = newHead
+	p.state = statedb
+
+	// Run the reorg between the old and new head and figure out which accounts
+	// need to be rechecked and which transactions need to be readded
+	if reinject, inclusions := p.reorg(oldHead, newHead); reinject != nil {
+		for addr, txs := range reinject {
+			// Blindly push all the lost transactions back into the pool
+			for _, tx := range txs {
+				p.reinject(addr, tx)
+			}
+			// Recheck the account's pooled transactions to drop included and
+			// invalidated one
+			p.recheck(addr, inclusions)
+		}
+	}
+	// Flush out any blobs from limbo that are older than the latest finality
+	p.limbo.finalize(p.chain.CurrentFinalBlock())
+
+	// Reset the price heap for the new set of basefee/blobfee pairs
+	var (
+		basefee = uint256.MustFromBig(misc.CalcBaseFee(p.chain.Config(), newHead))
+		blobfee = uint256.MustFromBig(big.NewInt(params.BlobTxMinDataGasprice))
+	)
+	if newHead.ExcessDataGas != nil {
+		blobfee = uint256.MustFromBig(misc.CalcBlobFee(*newHead.ExcessDataGas))
+	}
+	p.evict.reinit(basefee, blobfee, false)
+
+	basefeeGauge.Update(int64(basefee.Uint64()))
+	blobfeeGauge.Update(int64(blobfee.Uint64()))
+	p.updateStorageMetrics()
+}
+
+// reorg assembles all the transactors and missing transactions between an old
+// and new head to figure out which account's tx set needs to be rechecked and
+// which transactions need to be requeued.
+//
+// The transactionblock inclusion infos are also returned to allow tracking any
+// just-included blocks by block number in the limbo.
+func (p *BlobPool) reorg(oldHead, newHead *types.Header) (map[common.Address][]*types.Transaction, map[common.Hash]uint64) {
+	// If the pool was not yet initialized, don't do anything
+	if oldHead == nil {
+		return nil, nil
+	}
+	// If the reorg is too deep, avoid doing it (will happen during snap sync)
+	oldNum := oldHead.Number.Uint64()
+	newNum := newHead.Number.Uint64()
+
+	if depth := uint64(math.Abs(float64(oldNum) - float64(newNum))); depth > 64 {
```

Comment:
Are we sure we still need this restriction? Previously we use this to prevent some deep reorgs in proof-of-work which we believe is not possible to happen in mainnet. But I guess it might be possible in proof-of-stake in case of staying in a long-term-non-finalized network?

Reply:
The 64 is more like a hard cap on how much data we're willing to reshuffle. It would say it may be less probable in PoS than it was in PoW, but in both case it's unrelated to how likely it is, rather it's just a cap on how many blocks we'd reprocess. If the reorg goes deeper than that, we just don't want to txpool to spin too much. It will just drop the lost txs on the floor and let it be someone else's problem to resurrect them.